// sunflower-antd/es/useSearchResult/index.js
import { useState, useEffect } from 'react';
export var useSearchResult = function (_a) {
    var search = _a.search, _b = _a.autoFirstSearch, autoFirstSearch = _b === void 0 ? true : _b, defaultRequestData = _a.defaultRequestData;
    var _c = useState(), requestData = _c[0], setRequestData = _c[1];
    var _d = useState(), responseData = _d[0], setResponseData = _d[1];
    var _e = useState(false), loading = _e[0], setLoading = _e[1];
    var _f = useState(false), defaultRequestDataLoading = _f[0], setDefaultRequestDataLoading = _f[1];
    var searchFunc = function (data) {
        setRequestData(data);
        setLoading(true);
        return Promise.resolve(search && search(data)).then(function (response) {
            setResponseData(response);
            setLoading(false);
        });
    };
    useEffect(function () {
        setDefaultRequestDataLoading(true);
        var value;
        if (typeof defaultRequestData === 'function') {
            value = defaultRequestData();
        }
        else {
            value = defaultRequestData;
        }
        Promise.resolve(value).then(function (data) {
            setRequestData(data);
            setDefaultRequestDataLoading(false);
            if (autoFirstSearch) {
                searchFunc(data);
            }
        }).catch(function () {
            setDefaultRequestDataLoading(false);
        });
    }, []);
    return {
        loading: loading,
        requestData: requestData,
        setRequestData: setRequestData,
        responseData: responseData,
        defaultRequestDataLoading: defaultRequestDataLoading,
        search: searchFunc,
    };
};
